### PR TITLE
[`SwitchTransformer`] Significant performance improvement on MoE blocks

### PR DESCRIPTION
**Pull request by ranggihwang**
# What does this PR do?

**This is an edited version of the previously closed PR (https://github.com/huggingface/transformers/pull/30490)**

This PR includes a performant implementation of `SwitchTransformersSparseMLP` in the Google SwitchTransformer.
In the current implementation of the SwitchTransformer, it spans all possible experts, including the inactive ones.

```python
for idx, expert in enumerate(self.experts.values()):
            token_indices = router_mask[:, :, idx].bool()
            next_states[token_indices] = expert(hidden_states[token_indices]).to(next_states.dtype)
```

This results in serious performance degradation of the SwitchTransformer.

스크린샷 2024-04-26 오전 2 16 44
As shown in this figure, the current implementation of the SwitchTransformer spans inactive experts, unnecessarily increasing latency.

스크린샷 2024-04-26 오전 2 17 37
This issue can be particularly severe in models with a larger number of experts, as it needlessly spans more experts.

However, in my custom implementation of `SwitchTransformersSparseMLP`, it only accesses and computes the active experts.

**Advantages**
* This can significantly reduce the latency of the SwitchTransformer and make the model more accessible to a broader range of users.
* This change achieves greater latency reductions when expert parameters are offloaded to the CPU or SSD.
* This change addresses the problem of increasing latency proportional to the number of experts.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker and @younesbelkada

### CODE REVIEW

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot ! Looks very good ! Can you make sure the styling checks pas `make fixup` && `make fix-copies`

**Comment by ranggihwang** [comment]
Thanks @younesbelkada 
I've done `make fixup && make fix-copies` before in https://github.com/huggingface/transformers/pull/30490#pullrequestreview-2025011217
But I was requested to revert in the end.

How do I need to this correctly? Would you please let me know how to do this?

**Comment by younesbelkada** [comment]
Hi @ranggihwang  
Thanks ! Hmm I think there was a misunderstanding on my side at that time, if you could run the styling checks and push the results here (it should only change 2 files, switch transformers & gpt_san_japanese file), it would be great !

**Comment by younesbelkada** [comment]
since gpt_san_japanese uses blocks that are copied from switch transformers, running `make fix-copies` will propagate the changes you introduced in that file as well, see: https://app.circleci.com/pipelines/github/huggingface/transformers/94615/workflows/e1e9d110-614a-411a-a0f5-b7d4146e4db8/jobs/1241937

**Comment by younesbelkada** [comment]
Perfect thanks for the heads up @amyeroberts ! 
@ranggihwang feel free to proceed as suggested by amy 🙏

**Comment by ranggihwang** [comment]
Okay, now `make fixup` is done!

**Review by younesbelkada — APPROVED** [review]
Thanks !

**Comment by ArthurZucker** [comment]
I'll review this as I reviewed the previous PR, want to make sure the suggestions are all applied!

File: src/transformers/models/switch_transformers/modeling_switch_transformers.py
Comment (ArthurZucker) [inline src/transformers/models/switch_transformers/modeling_switch_transformers.py:305]:
```suggestion
idx_mask = router_mask.reshape(batch*seq_len, num_experts).transpose(0,1).sum(dim=1) 
idx_mask = torch.nonzero(idx_mask, as_tuple=True)[0].tolist()
```
+ the comment about shapes! 🤗

**Review by ArthurZucker — COMMENTED** [review]
Could you apply the suggestion I did in the previous PR

File: src/transformers/models/switch_transformers/modeling_switch_transformers.py
Comment (ranggihwang) [reply src/transformers/models/switch_transformers/modeling_switch_transformers.py:305]: The batch_size, seq_len, and num_experts are not defined in the funciton.
So, I've defined it with the router_mask and reflected your suggestions.

Thank you @ArthurZucker !

**Comment by ranggihwang** [comment]
@ArthurZucker @younesbelkada 
Please let me know if I need to rebase it again :)

**Review by younesbelkada — APPROVED** [review]
Still LGTM ! Let's wait for @ArthurZucker 's final review!

File: src/transformers/models/switch_transformers/modeling_switch_transformers.py
Comment (ArthurZucker) [reply src/transformers/models/switch_transformers/modeling_switch_transformers.py]: thanks!

**Review by ArthurZucker — APPROVED** [review]
Thanks a lot! 🤗

**Comment by ArthurZucker** [comment]
Could this be propagated to the qwen code @ranggihwang ? I know that they have some variants with lots of experts!

**Comment by ranggihwang** [comment]
@ArthurZucker  I think it can be adopted for many MoE models in HuggingFace not only qwen-moe but also for NLLB-MoE, Mixtral, etc.

**Comment by ArthurZucker** [comment]
awesome! Then if you are interested feel free to open a PR and ping me! 🤗

**Pull request merged**
